### PR TITLE
Add SPIFFE ID character encoding restrictions

### PR DESCRIPTION
**Pull request by justinburke**
This commit adds SPIFFE ID restrictions discussed in SIG-Spec.
Specifically, it forbids percent encoding, requires that scheme and host
components be lowercased, and forbids relative path modifiers and
trailing slashes in the path component.

A subsequent commit will add an appendix item that describes how a
SPIFFE ID can be validated in a lightweight way.

Signed-off-by: Evan Gilman <egilman@vmware.com>

### CODE REVIEW

File: standards/SPIFFE-ID.md [hunk 1]
```diff
@@ -57,9 +64,14 @@ When a collision does occur, those trust domains will continue to operate indepe
 
 
 ### 2.2. Path
-The path component of a SPIFFE ID allows for the unique identification of a given workload. The meaning behind the path is left open ended and the responsibility of the administrator to define.
+The path component of a SPIFFE ID allows for the unique identification of a given workload. The meaning behind the path is left open ended and is the responsibility of the administrator to define.
+
+Valid SPIFFE ID path components adhere to the following rules:
+* The path component MUST NOT include percent-encoded characters.
+* The path component MUST NOT include empty path segments or relative path modifiers (i.e. `//`, `/./`, or `/../`)
```
Comment (justinburke) [inline standards/SPIFFE-ID.md]: (i.e. `.`, `..`)

Comment (azdagron) [reply standards/SPIFFE-ID.md]:
```suggestion
* The path component MUST NOT include segments that are empty or are relative path modifiers (i.e. `.`, `..`)
```

File: standards/SPIFFE-ID.md [hunk 2]
```diff
@@ -38,14 +38,21 @@ In order to communicate an identity, we must first define an identity namespace.
...
 
 ### 2.1. Trust Domain
 The trust domain corresponds to the trust root of a system. A trust domain could represent an individual, organization, environment or department running their own independent SPIFFE infrastructure.
 
 Trust domain names are nominally self-registered, unlike public DNS there is no delegating authority that acts to assert and register a base domain name to an actual legal real-world entity, or assert that legal entity has fair and due rights to any particular trust domain name.
 
-The trust domain name is defined as the authority component of the URI - specifically, the `host` part of the authority. The `userinfo` and `port` parts of the authority component MUST NOT be set, and the `:` delimiter MUST NOT be present. Please see section 3.2 of [RFC 3986](https://tools.ietf.org/html/rfc3986) for more information.
+The trust domain name is defined as the authority component of the URI with the following restrictions applied:
+* The `host` part of the authority MUST be set.
+* The `userinfo` and `port` parts of the authority component MUST NOT be set.
```
Comment (justinburke) [inline standards/SPIFFE-ID.md]:
```suggestion
* The `host` part of the authority MUST NOT be empty.
* The `userinfo` and `port` parts of the authority component MUST be empty.
```

File: standards/SPIFFE-ID.md [hunk 3]
```diff
@@ -38,14 +38,21 @@ In order to communicate an identity, we must first define an identity namespace.
...
+* The `host` part of the authority MUST be lowercase.
+* The `host` part of the authority MUST contain only letters, numbers, dots, and dashes ([a-z1-9.-]).
```
Comment (justinburke) [inline standards/SPIFFE-ID.md]:
```suggestion
* The `host` part of the authority MUST contain only letters, numbers, dots, and dashes ([a-z0-9.-]).
```

File: standards/SPIFFE-ID.md [hunk 4]
```diff
@@ -57,9 +64,14 @@ When a collision does occur, those trust domains will continue to operate indepe
 
 
 ### 2.2. Path
-The path component of a SPIFFE ID allows for the unique identification of a given workload. The meaning behind the path is left open ended and the responsibility of the administrator to define.
+The path component of a SPIFFE ID allows for the unique identification of a given workload. The meaning behind the path is left open ended and is the responsibility of the administrator to define.
```
Comment (justinburke) [inline standards/SPIFFE-ID.md]:
```suggestion
The path component of a SPIFFE ID allows for the unique identification of a given workload. The meaning behind the path is left open-ended and is the responsibility of the administrator to define.
```

**Review by spikecurtis — APPROVED** [review]

**Comment by justinburke** [comment]
"Pull request authors can't approve their own pull request".....

This PR looks good to me.

:ship it:

**Comment by justinburke** [comment]
BTW, @evan2645, since GitHub doesn't consider you the pull request author, you have the power to approve this request. That... seems kinda broken...

**Pull request merged**